crates/executor/state_handler/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+//! A state handler that can be used to store and retrieve state.
+mod snapshot_state;
+
+use std::ops::Add;
+
+use crate::snapshot_state::{Snapshot, SnapshotState};
+use allocator_api2::alloc::{Allocator, Global};
+use allocator_api2::vec::Vec;
+use ixc_account_manager::state_handler::std::{StdStateError, StdStateManager};
+use ixc_message_api::AccountID;
+
+/// A store that can be used to store and retrieve state.
+pub trait Store {
+    /// Get the value for the given key.
+    fn get<A: Allocator>(&self, key: &Vec<u8>, allocator: A) -> Option<Vec<u8, A>>;
+}
+
+/// StateHandler is a cache-based state handler that can be used to store and retrieve state.
+pub struct StateHandler<S: Store> {
+    snapshot_state: SnapshotState<S>,
+    /// Checkpoints are used to revert state changes.
+    /// checkpoints are used to follow the call stack of a transaction
+    /// and revert the state changes when the transaction is rolled back.
+    checkpoints: Vec<Snapshot>,
+}
+
+impl<S: Store> StateHandler<S> {
+    /// Create a new state handler with the given store.
+    pub fn new(store: S) -> Self {
+        Self {
+            snapshot_state: SnapshotState::new(store),
+            checkpoints: Vec::with_capacity_in(200, Global),
+        }
+    }
+
+    pub(crate) fn construct_key(
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        accumulator: bool,
+    ) -> Vec<u8> {
+        let separator: [u8; 1] = [b'/'];
+        match scope {
+            // account / 0 / key
+            // account / 1 / scope / key
+            // account / 2 (accumulator) / key
+            // account / 3 (scoped accumulator) / scope / key
+            Some(scope) => {
+                let ac = account_id.to_le_bytes();
+                let sc = scope.to_le_bytes();
+                let mut new_key = Vec::new_in(Global);
+                new_key.extend_from_slice(&ac);
+                new_key.extend_from_slice(&separator);
```

Comment:
I don't think we need this separator at all, instead the prefix 0-3 serves as the separator